### ISSUE TITLE
Stack buffer overflow

### ISSUE THREAD

**Issue by ZOTTCE** [amethyst/laminar#272]
There is a critical bug with `CongestionHandler`.
You can easily get `STATUS_STACK_BUFFER_OVERRUN` error on Windows with like 3000 virtual connections. Also this can be reproduced on Linux servers. Today I have bad feedback because my application causes crashes on real working servers.

This test is able to show what I'm talking about
```Rust
    #[test]
    fn billion_virtual_connections() {
        let mut connections = Vec::with_capacity(3000);
        
        for i in 0..connections.capacity() {
            connections.push(create_virtual_connection());
        }
    }
```

**Comment by BourgondAries** [comment]
`create_virtual_connection` is unspecified and does not exist in laminar. Please provide it.

**Comment by BourgondAries** [comment]
Works just fine on linux. I increased it to 30k and it just consumes a lot of memory. Are you running out of memory? Is the OOM killer invoked? On my machine it consume somewhere around 5-6 GB of RAM.

**Comment by ZOTTCE** [comment]
Yeah, I see it's really consumes so much memory. I don't think it's an OOM killer. I'm stuck with 32 system because of a main application (my app is a dynamic library).
Is it fine to just disable this handler?

**Comment by BourgondAries** [comment]
If you're on a 32-bit system then you could address at most 4 GB, maybe more with extended addressing, can you check if it's OOM? Not sure what you mean by "disable this handler".

**Comment by ZOTTCE** [comment]
I mean can I replace current implementation with empty functions like this? Will it affect something internal?

```Rust
pub struct CongestionHandler {}

impl CongestionHandler {
    pub fn new(config: &Config) -> CongestionHandler {
        CongestionHandler {}
    }

    pub fn process_incoming(&mut self, _incoming_seq: u16) {}

    pub fn process_outgoing(&mut self, _seq: u16, _time: Instant) {}
}
```

**Comment by BourgondAries** [comment]
Where are you getting the idea from that it's the congestion handler? Can you check the kernel log for OOM messages?

**Comment by ZOTTCE** [comment]
I got a crash log with backtrace and it looks like this:
```
std::alloc::rust_oom
alloc::alloc::handle_alloc_error
< ... some internal memory allocations ..  >
SequenceBuffer::with_capacity
CongestionHandler::new
```
And what about the kernel log I'll talk to the server administrator.

**Comment by BourgondAries** [comment]
That indeed looks like OOM. You could try editing the capacity of the internal sequencebuffer to be smaller. Then each sequence buffer can grow as demand requires it. This will incur some more allocations during run-time though. Currently we do

```
congestion_data: SequenceBuffer::with_capacity(<u16>::max_value()),
```

Maybe that data structure needs to be changed to a HashSet instead since 65536 elements of type `CongestionData` which contains `Instant` and `SequenceNumber`, on unix platforms this is `64 + 16 bits` so 10 bytes in total, meaning each congestion handler allocates ~655 kB which is quite a lot.

So we can try to make the capacity lower, or replace the functionality by a HashMap or HashSet.

**Comment by ZOTTCE** [comment]
It's about ~766 kB for each connection, because `SequenceBuffer` does two allocations. These allocations exist even the connection isn't active, so ... It is not so hard to drain all memory in server (especially for 32 bit systems), make a small program using laminar and try to connect to server and done the server is dead.

So, the questions is can I remove this `CongestionHandler` while you're fixing this? Because as I said there is no usage of the handler.

**Comment by BourgondAries** [comment]
I'm not fixing this since I don't need it. You can try it yourself. Try removing it and check if all tests succeed. It would be appreciated if you could make a pull request for any fix you find.

**Comment by moien007** [comment]
Please mention `32bit` in the title.
